File: Solved/problem658.c
//
//  problem658.c
//
//  Created by Dominique Normand on 3/2/19.
//

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define ulong unsigned long

const ulong modulo     = 1000000007;
const ulong LETTERS    = 1E5;
const ulong MAX_LENGTH = 1E12;

ulong modPow(ulong value, ulong exp)
{
    ulong r     = 1;
    ulong base  = value;
    if (base >= modulo)
        base %= modulo;

    if (base == 0)
        return 0;

    while (exp > 0)
    {
        if ((exp & 1) == 1)
            r = (r * base) % modulo;

        exp >>= 1;
        base = (base * base) % modulo;
    }

    return r;
}

long modInv(long newR)
{
    long t    = 0;
    long newT = 1;
    long r    = modulo;
    long q, lastT, lastR;

    while (newR != 0)
    {
        q = r / newR;
        lastT = t;
        lastR = r;
        t = newT;
        r = newR;
        newT = lastT - q * newT;
        newR = lastR - q * newR;
    }
    if (r != 1)
    {
        printf("failed to process modInv - values not co-prime\n");
        exit(-1);
    }

    if (t < 0)
        t += modulo;
    
    if (modulo < 0)
        return -t;
    return t;
};

ulong A(ulong l, ulong length)
{
    if (l == 0)
        return 1;
    if (l == 1) // avoid dividing by 0
        return (length + 1) % modulo;
    ulong total = modPow(l, length + 1) - 1;
    if (total < 0)
        total += modulo;
    long divisor = modInv(l-1);
    total = (total * divisor) % modulo;
    
    return total;
}

ulong I(ulong letters, ulong length, ulong* factors, int trace)
{
    ulong total = 0;

    for (ulong i = 1, j = letters-1; i <= letters; i++, j--)
    {
        if (trace)
        {
            fprintf(stdout, "\r%lu", i);
            fflush(stdout);
        }
        
        ulong f = A(j, length);
        ulong x = factors[j];
        int negative = (x < 0);
        if (negative)
            x = -x;
        
        ulong y = (f * x) % modulo;
        
        if (negative)
        {
            if (total < y)
                total += modulo;
            
            total = (total - y) % modulo;
        }
        else
        {
            total = (total + y) % modulo;
        }
    }
    
    return total;
}

void noTrace(ulong letter) {}

static int count = 0;

void Trace(ulong letter)
{
    if (count == 0)
    {
        fprintf(stdout, "\r%lu", letter);
        fflush(stdout);
    }
    if (count++ >= 999)
        count = 0;
}

ulong Solve(ulong letters, ulong length, int trace)
{
    ulong total;
    
    void (*doTrace)(ulong) = trace ? Trace : noTrace;
    
    ulong *previousRow = (ulong*)malloc(sizeof(ulong) * letters);
    ulong *currentRow  = (ulong*)malloc(sizeof(ulong) * letters);
    ulong *factors     = (ulong*)malloc(sizeof(ulong) * letters);
    
    for (ulong letter = 1; letter <= letters; letter++)
    {
        doTrace(letter);
        
        ulong max = letters-letter;
        
        if (letter == 1)
        {
            total = 0;
            for (ulong l = 0; l <= max; l++)
            {
                ulong v = (l & 1) ? modulo-1 : 1;
                currentRow[l] = v;
                total += v;
            }
            total %= modulo;
        }
        else // use previous row to figure it out
        {
            currentRow[0] = letter;
            total  = letter;
            for (ulong l = 1; l <= max; l++)
            {
                ulong v = previousRow[l];
                ulong v2= currentRow[l-1];
                if (v >= v2)
                    v -= v2;
                else
                    v = v + modulo - v2;
                
                currentRow[l] = v;
                
                total += v;
            }
            total %= modulo;
        }
        factors[letter-1] = total;
        
        // Switch rows to avoid allocating memory
        ulong* r = currentRow;
        currentRow = previousRow;
        previousRow = r;
    }
    
    free(previousRow);
    free(currentRow);
    
    if (trace)
    {
        printf("\nStep 1 done\n");
        printf("Last step - Consolidation\n");
    }
    
    ulong result = I(letters, length, factors, trace);
    
    free(factors);
    
    if (trace)
        printf("\nDone\n");
    
    return result;
}

void Assert(ulong value, ulong expected)
{
    if (value == expected)
        return;
    printf("Test failed: got %lu but expected %lu\n", value, expected);
    exit(-1);
}

int main(int argc, const char * argv[])
{
    Assert(Solve(4, 4, 0), 406);
    Assert(Solve(8, 8, 0), 27902680);
    Assert(Solve(10,100,0), 983602076);
    
    ulong start = clock();
    ulong answer = Solve(LETTERS, MAX_LENGTH, 1);
    ulong end = clock();
    
    ulong total_t = (ulong )(end - start) / CLOCKS_PER_SEC;
    
    printf("Answer is %lu executed in %lu seconds\n", answer, total_t);
    return 0;
}

File: Solved/problem658.js
const assert = require('assert');
const prettyTime= require("pretty-hrtime");
const announce = require('@dn0rmand/project-euler-tools/src/announce');

require('@dn0rmand/project-euler-tools/src/numberHelper');

const modulo     = 1000000007;

const LETTERS    = 1E5;
const MAX_LENGTH = 1E12;

class problem658
{
    constructor(letters, length, trace)
    {
        this.modulo      = 1000000007;
        this.factors     = new Uint32Array(letters);
        this.currentRow  = new Uint32Array(letters);
        this.previousRow = new Uint32Array(letters);
        this.letters     = Number(letters);
        this.length      = Number(length);
        this.trace       = trace === true;
    }

    A(l)
    {
        if (l == 0)
            return 1;
        if (l == 1) // avoid dividing by 0
            return (this.length + 1) % this.modulo;

        let total = l.modPow(this.length + 1, this.modulo) - 1;
        if (total < 0)
            total += this.modulo;
        total = total.modDiv(l-1, this.modulo);

        return total;
    }

    I()
    {
        let total = 0;

        for (let i = 1, j = this.letters-1; i <= this.letters; i++, j--)
        {
            if (this.trace)
                process.stdout.write(`\r${i}`);

            let f = this.A(j);

            let x = this.factors[j] || 0;
            let negative = (x < 0);
            if (negative)
                x = -x;

            let y = f.modMul(x, this.modulo);

            if (negative)
            {
                if (total < y)
                    total += this.modulo;

                total = (total - y) % this.modulo;
            }
            else
            {
                total = (total + y) % this.modulo;
            }
        }

        return total;
    }

    I2(letter)
    {
        let total = 0;
        let max = this.letters-letter;

        if (letter === 1)
        {
            for (let l = 0; l <= max; l++)
            {
                let v = (l & 1) ? this.modulo-1 : 1;
                this.currentRow[l] = v;
                total += v;
                if (total >= this.modulo)
                    total -= this.modulo;
            }
        }
        else // use previous row to figure it out
        {
            this.currentRow[0] = letter;
            total  = letter;
            for (let l = 1; l <= max; l++)
            {
                let v = this.previousRow[l];
                let v2 = this.currentRow[l-1];
                if (v >= v2)
                    v -= v2;
                else
                    v = v + this.modulo - v2;

                this.currentRow[l] = v;

                total += v;

                if (total >= this.modulo)
                    total -= this.modulo;
            }
        }
        this.factors[letter-1] = total;
    }

    S()
    {
        let timer = process.hrtime();
        let count = 0;
        let total;
        let max;
        let doTrace = (l) => {

        };
        if (this.trace)
        {
            doTrace = (letter) => {
                if (count === 0)
                    process.stdout.write(`\r${letter}`);
                if (count++ >= 999)
                    count = 0;
            }
        }

        for (let letter = 1; letter <= this.letters; letter++)
        {
            doTrace(letter);

            max = this.letters-letter;

            if (letter === 1)
            {
                total = 0;
                for (let l = 0; l <= max; l++)
                {
                    let v = (l & 1) ? this.modulo-1 : 1;
                    this.currentRow[l] = v;
                    total += v;
                    if (total >= this.modulo)
                        total -= this.modulo;
                }
            }
            else // use previous row to figure it out
            {
                this.currentRow[0] = letter;
                total  = letter;
                for (let l = 1; l <= max; l++)
                {
                    let v = this.previousRow[l];
                    let v2 = this.currentRow[l-1];
                    if (v >= v2)
                        v -= v2;
                    else
                        v = v + this.modulo - v2;

                    this.currentRow[l] = v;

                    total += v;

                    if (total >= this.modulo)
                        total -= this.modulo;
                }
            }
            this.factors[letter-1] = total;

            // this.I2(letter);

            // Switch rows to avoid allocating memory
            let r = this.currentRow;
            this.currentRow  = this.previousRow;
            this.previousRow = r;
        }
        timer = process.hrtime(timer);

        if (this.trace)
        {
            console.log("\rStep 1 done in", prettyTime(timer, {verbose:true}));
            console.log("Last step - Consolidation");
        }

        let result = this.I();

        if (this.trace)
            console.log('\rDone     ');

        return result;
    }

    static Solve(letters, length, trace)
    {
        let p = new problem658(letters, length, trace);
        return p.S();
    }
}

assert.equal(problem658.Solve(4,4), 406);
assert.equal(problem658.Solve(8,8), 27902680);
assert.equal(problem658.Solve(10,100), 983602076);

let timer = process.hrtime();
let answer = problem658.Solve(LETTERS, MAX_LENGTH, true);
timer = process.hrtime(timer);

timer = prettyTime(timer, {verbose:true});
console.log('Answer is', answer, 'calculated in', timer);
// announce(658, `Answer is ${answer} calculated in ${ timer}`);
console.log('Done');